File: tests/cohendiv.c
#include <stdio.h>
#include <stdlib.h>

void vassume(int b){}
void vtrace1(int q, int r, int a, int b, int x, int y){}
void vtrace2(int q, int r, int a, int b, int x, int y){}
void vtrace3(int q, int r, int x, int y){}

int mainQ(int x, int y){
    vassume(x >= 1 && y >= 1);
    
    int q=0;
    int r=x;
    int a=0;
    int b=0;
    while(1) {
	vtrace1(q, r, a, b, x, y);
	if(!(r>=y))
	    break;
	a=1;
	b=y;
	  
	while (1){
	    vtrace2(q, r, a, b, x, y);
	    if(!(r >= 2*b))
		break;
	       
	    a = 2*a;
	    b = 2*b;
	}
	r=r-b;
	q=q+a;
    }
    vtrace3(q, r,x, y);
    return q;
}

void main(int argc, char **argv){
    mainQ(atoi(argv[1]), atoi(argv[2]));
}

File: benchmark/java/nla_full/CohenDiv.java
public class CohenDiv {
    public static void vtrace1(int q, int r, int a, int b, int x, int y){}
    public static void vtrace2(int q, int r, int a, int b, int x, int y){}
    public static void vtrace3(int x, int y, int q, int r){}

    public static void main (String[] args) {}
     

    public static int mainQ(int x, int y) {
	//assert(y >= 1);
	assert(x >= 1);
	assert(y >= 1);

	int q=0;
	int r=x;
	int a=0;
	int b=0;
	  
	while(true) {
	    //assert(q*y + r == x);
	    //assert(a*y == b);
	    
	    vtrace1(q,r,a,b,x,y);
	    
	    if(!(r>=y)) break;
	    a=1;
	    b=y;

	    while (true) {
		//assert(q*y + r == x);
		//assert(a*y == b);

		vtrace2(q,r,a,b,x,y);
		if(!(r >= 2*b)) break;

		a = 2*a;
		b = 2*b;
	    }
	    r=r-b;
	    q=q+a;
	}
	//assert(q*y + r == x);
	vtrace3(x,y,q,r);
	return q;
    }
}
